Extension/src/LanguageServer/settings.ts

```diff
@@ -25,146 +26,157 @@ nls.config({ messageFormat: nls.MessageFormat.bundle, bundleFormat: nls.BundleFo
 const localize: nls.LocalizeFunc = nls.loadMessageBundle();
 
 export interface Excludes {
-    [key: string]: boolean | { when: string };
+    [key: string]: (boolean | { when: string });
 }
 
+export interface Associations {
+    [key: string]: string;
+}
+
+// Settings that can be undefined have default values assigned in the native code or are meant to return undefined.
 export interface WorkspaceFolderSettingsParams {
     uri: string | undefined;
-    intelliSenseEngine: string | undefined;
-    intelliSenseEngineFallback: boolean | undefined;
-    autocomplete: string | undefined;
-    autocompleteAddParentheses: boolean | undefined;
-    errorSquiggles: string | undefined;
-    exclusionPolicy: string | undefined;
-    preferredPathSeparator: string | undefined;
+    intelliSenseEngine: string;
+    intelliSenseEngineFallback: boolean;
+    autocomplete: string;
+    autocompleteAddParentheses: boolean;
+    errorSquiggles: string;
+    exclusionPolicy: string;
+    preferredPathSeparator: string;
     intelliSenseCachePath: string | undefined;
```

Comment:
intelliSenseCachePath does not seem like it should have the `| undefined` type. It is never assigned an undefined.

Reply:
`util.resolveCachePath(settings.intelliSenseCachePath` converts any undefined value in `settings.intelliSenseCachePath` to an empty string and `settings.intelliSenseCachePath` converts any empty string to an undefined value, so it's a little confusing.

Reply:
The same issue occurs with clangFormatPath

Reply:
clangTidyPath is also in the same boat.  I fixed them all.